### PR TITLE
Plfm 7444

### PR DESCRIPTION
**Pull request by SandhraSokhal**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java [hunk 1]
```diff
@@ -112,7 +116,23 @@ public class DBOFileHandleDaoImpl implements FileHandleDao {
...
 			+ " AND T." + COL_FILES_ID + "= ?";
-	
+
+	private static final String SELECT_FILE_SUMMARY_FOR_ID_AND_VERSION = "SELECT COUNT(*) AS COUNT, " +
+			"MD5(GROUP_CONCAT(F." + COL_FILES_CONTENT_MD5 + " ORDER BY F." + COL_FILES_CONTENT_MD5 + " ASC )) AS CHECKSUM, " +
+			"SUM(F." + COL_FILES_CONTENT_SIZE + ") AS SIZE " +
+			" FROM " + TABLE_REVISION + " R JOIN " + TABLE_FILES + " F ON R." + COL_REVISION_FILE_HANDLE_ID + " = F." + COL_FILES_ID +
+			" WHERE (R." + COL_REVISION_OWNER_NODE + ", R." + COL_REVISION_NUMBER + ") IN (:pairs)";
+
+	private static final RowMapper<FileSummary> FILE_SUMMARY_ROW_MAPPER = new RowMapper<FileSummary>() {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java]: make this a lambda function

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java [hunk 2]
```diff
@@ -594,4 +614,8 @@ public void truncateTable() {
 		}
 	}
 
+	@Override
+	public FileSummary getFileSummary(Map<String, List<Long[]>> namedParameters) {
+		return namedJdbcTemplate.query(SELECT_FILE_SUMMARY_FOR_ID_AND_VERSION, namedParameters, FILE_SUMMARY_ROW_MAPPER).get(0);
+	}
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java]: qureyForObject()

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java [hunk 3]
```diff
@@ -594,4 +614,8 @@ public void truncateTable() {
 		}
 	}
 
+	@Override
+	public FileSummary getFileSummary(Map<String, List<Long[]>> namedParameters) {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImpl.java]: The caller should not know anything about a query or the parameters.  We can pass the List<EntityRef> as a param.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/FileSummary.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return "FileSummary{" +
+                "checksum='" + checksum + '\'' +
+                ", size=" + size +
+                ", count=" + count +
+                '}';
+    }
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/FileSummary.java]: hash and equals?

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/table/Dataset.json [hunk 5]
```diff
@@ -12,6 +12,14 @@
 			"items": {
 				"$ref": "org.sagebionetworks.repo.model.EntityRef"
 			}
+		},
+		"size": {
+			"description": "The cumulative size, in bytes, of all items in the dataSet.",
+			"type": "integer"
+		},
+		"checksum": {
+			"description": "The checksum is computed over a sorted concatenation of the checksums of all items in the dataSet.",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/table/Dataset.json]: might need the pseudo-code to show how we compute the checksum.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/EntityManagerImpl.java [hunk 6]
```diff
@@ -124,6 +128,18 @@ public <T extends Entity> String createEntity(UserInfo userInfo, T newEntity, St
...
 
+	void calculateSizeAndChecksum(Dataset dataset) {
+		if (dataset.getItems() == null || dataset.getItems().isEmpty()) {
+			return;
+		}
+		FileSummary fileSummary = fileHandleManager.getFileSummary((dataset).getItems());
+		/*if ((dataset).getItems().size() != fileSummary.getCount()) {
+			throw new IllegalArgumentException("Invalid item present for dataset.");
+		}*/
+		(dataset).setChecksum(fileSummary.getChecksum());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/EntityManagerImpl.java]: no need to cast here.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/EntityManagerImpl.java [hunk 7]
```diff
@@ -348,6 +364,10 @@ public <T extends Entity> boolean updateEntity(UserInfo userInfo, T updated, boo
 			newVersion = false;
 		}
 
+		if (updated instanceof Dataset) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/EntityManagerImpl.java]: will not be needed.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java [hunk 8]
```diff
@@ -1392,4 +1395,19 @@ public boolean isMatchingMD5(String sourceFileHandleId, String targetFileHandleI
 		
 		return fileHandleDao.isMatchingMD5(sourceFileHandleId, targetFileHandleId);
 	}
+
+	@Override
+	public FileSummary getFileSummary(final List<EntityRef> entityRefs) {
+		List<Long[]> specificIdVersionPairs = new ArrayList<>(entityRefs.size());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java]: 1401 through 1409 can move to the dao.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java [hunk 9]
```diff
@@ -1392,4 +1395,19 @@ public boolean isMatchingMD5(String sourceFileHandleId, String targetFileHandleI
 		
 		return fileHandleDao.isMatchingMD5(sourceFileHandleId, targetFileHandleId);
 	}
+
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java]: After discussion we all think this would fit better in NodeManager and NodeDao.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/EntityManagerImplUnitTest.java [hunk 10]
```diff
@@ -96,11 +64,46 @@
 import org.sagebionetworks.repo.model.schema.ListValidationResultsResponse;
 import org.sagebionetworks.repo.model.schema.ValidationResults;
 import org.sagebionetworks.repo.model.schema.ValidationSummaryStatistics;
+import org.sagebionetworks.repo.model.table.Dataset;
 import org.sagebionetworks.repo.model.table.Table;
 import org.sagebionetworks.repo.web.NotFoundException;
 
-import com.google.common.collect.Lists;
-import com.google.common.collect.Sets;
+import java.util.ArrayList;
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/EntityManagerImplUnitTest.java]: can also be moved to the DatasetMetadataProvider unit test.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/EntityManagerImplUnitTest.java [hunk 11]
```diff
@@ -1675,6 +1678,52 @@ public void testGetAnnotationsWithUnauthorized() {
...
+
+		FileSummary expectedSummary = new FileSummary("gef4563h", 15L, 1);
+		ArgumentCaptor<List<EntityRef>> captor = ArgumentCaptor.forClass(List.class);
+		when(mockFileHandleManager.getFileSummary(any())).thenReturn(expectedSummary);
+
+		//call under test
+		entityManager.calculateSizeAndChecksum(dataSet);
+		assertEquals(expectedSummary.getSize(), dataSet.getSize());
+		assertEquals(expectedSummary.getChecksum(), dataSet.getChecksum());
+		verify(mockFileHandleManager, times(1)).getFileSummary(captor.capture());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/EntityManagerImplUnitTest.java]: The captor is not used.
shortcut verify().getFileSummary(entityRefs)

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImplTest.java [hunk 12]
```diff
@@ -2619,4 +2620,26 @@ public void testUploadLocalFile() throws IOException, AmazonServiceException, Am
 		}
 		
 	}
+
+	@Test
+	public void testGetFileSummary() {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImplTest.java]: the new manager test will be only the doa was called with the same list.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/file/preview/StubFileMetadataDao.java [hunk 13]
```diff
@@ -1,25 +1,25 @@
...
 import org.sagebionetworks.repo.model.file.FileHandle;
 import org.sagebionetworks.repo.model.file.FileHandleResults;
 import org.sagebionetworks.repo.model.file.FileHandleStatus;
 import org.sagebionetworks.repo.web.NotFoundException;
 
-import com.google.common.collect.Maps;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/file/preview/StubFileMetadataDao.java]: this can also be removed.

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/EntityServiceImplAutowiredTest.java [hunk 14]
```diff
@@ -729,4 +727,56 @@ public void testUpdateTableWithSearchEnabled() {
...
+		toDelete.add(file2.getId());
+
+		final EntityRef entityRef2 = new EntityRef();
+		entityRef2.setEntityId(file2.getId());
+		entityRef2.setVersionNumber(file2.getVersionNumber());
+		createdDataset.getItems().add(entityRef2);
+		//call under test
+		final Dataset updatedDataset = entityService.updateEntity(adminUserId, createdDataset, false, null);
+		assertNotNull(updatedDataset);
+		assertNotNull(updatedDataset.getChecksum());
```
Comment (john-hill) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/EntityServiceImplAutowiredTest.java:778]: This check will  happen in the DAO test:
need to check that the checksum:
1: order the md5s of the two files
2. put into a comma separated string
3. calculate md5 of concat.

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/EntityServiceImplAutowiredTest.java [hunk 15]
```diff
@@ -729,4 +727,56 @@ public void testUpdateTableWithSearchEnabled() {
 		assertTrue(table.getIsSearchEnabled());
 		
 	}
+
+	@Test
+	public void testDatasetCreationAndUpdateCalculateSizeAndChecksum() {
```
Comment (john-hill) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/EntityServiceImplAutowiredTest.java:732]: this whole test moves to the dao test and test the actual MD5 (see MD5ChecksumHelper).

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 16]
```diff
@@ -5267,4 +5268,43 @@ public void testGetNodeNameWithNotFound() {
...
+	@Test
+	public void testFileSummary() {
+		//create a project
+		Node parent = privateCreateNew("parent");
+		parent.setNodeType(EntityType.project);
+		String parentId = nodeDao.createNew(parent);
+		toDelete.add(parentId);
+
+		// create a file 1
+		Node file1 = privateCreateNew("file1");
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java:5279]: file1 does not have a parent.  We can either remove the project or set the parentId of each file.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 17]
```diff
@@ -5267,4 +5268,43 @@ public void testGetNodeNameWithNotFound() {
...
+		String concatenatedString = String.join(",", md5List);
+		//calculate md5 of concatenated string
+		String expectedChecksum = DigestUtils.md5Hex(concatenatedString);
+		FileSummary expectedFileSummary = new FileSummary(expectedChecksum, expectedTotalSize, 2);
+
+		FileSummary fileSummary = nodeDao.getFileSummary(Arrays.asList(new EntityRef().setEntityId(file1Id).setVersionNumber(1L),
+				new EntityRef().setEntityId(file2Id).setVersionNumber(1L)));
+		assertNotNull(fileSummary);
+		assertEquals(expectedFileSummary, fileSummary);
+	}
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java:5308]: Need a test for the case where there list is empty.  When it is empty size=0, count=0, and md5=null.
Since the code ignores null entity Id then we need a test for the null case.

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/table/Dataset.json [hunk 18]
```diff
@@ -12,6 +12,14 @@
 			"items": {
 				"$ref": "org.sagebionetworks.repo.model.EntityRef"
 			}
+		},
+		"size": {
+			"description": "The cumulative size, in bytes, of all items(files) in the dataSet.",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/table/Dataset.json]: Add this is computed automatically by the service.  Same with the checksum we do not want users to think they need to provide these values.

**Review by john-hill — APPROVED** [review]

**Review by john-hill — CHANGES_REQUESTED** [review]
We need handle the empty items case or it will be a regression.

**Review by john-hill — APPROVED** [review]
